File: src/justitone/jidi/JidiEvent.java
package justitone.jidi;

import java.util.List;

import justitone.TokenPos;

public abstract class JidiEvent {
    public long tick;
    
    public static class NoteOn extends JidiEvent {
        @Override
        public String toString() {
            return tick + ": NoteOn";
        }

        NoteOn(long tick) {
            this.tick = tick;
        }
    }
    
    public static class NoteOff extends JidiEvent {
        @Override
        public String toString() {
            return tick + ": NoteOff";
        }

        NoteOff(long tick) {
            this.tick = tick;
        }
    }
    
    public static class Pitch extends JidiEvent {
        @Override
        public String toString() {
            return tick + ": Pitch [freq=" + freq + "]";
        }

        public float freq;

        Pitch(long tick, float freq) {
            this.tick = tick;
            this.freq = freq;
        }
    }
    
    public static class Instrument extends JidiEvent {
        @Override
        public String toString() {
            return tick + ": Instrument [instrument=" + instrument + "]";
        }
        
        public int instrument;
        
        Instrument(long tick, int instrument) {
            this.tick = tick;
            this.instrument = instrument;
        }
    }
    
    public static class Token extends JidiEvent {
        @Override
        public String toString() {
            return tick + ": Tokens";
        }
        
        public List<TokenPos> tokens;
        
        Token(long tick, List<TokenPos> tokens) {
            this.tick = tick;
            this.tokens = tokens;
        }
        
        public int start() {
            return tokens.stream().mapToInt(t -> t.start).min().orElse(0);
        }
        
        public int stop() {
            return tokens.stream().mapToInt(t -> t.stop).max().orElse(0);
        }
    }
}

File: src/main/kotlin/justitone/jidi/JidiEvent.kt
package justitone.jidi

import justitone.TokenPos

sealed class JidiEvent {
    abstract val tick: Long

    data class NoteOn(override val tick: Long) : JidiEvent()
    data class NoteOff(override val tick: Long) : JidiEvent()
    data class Pitch(override val tick: Long, val freq: Float) : JidiEvent()
    data class Instrument(override val tick: Long, val instrument: Int) : JidiEvent()
    data class Token(override val tick: Long, val tokens: List<TokenPos>) : JidiEvent() {
        val start by lazy { tokens.map { it.start }.min() ?: 0 }
        val stop by lazy { tokens.map { it.stop }.max() ?: 0 }
    }
}
